### ISSUE TITLE
Not compatible with the new version of Google Sheets?

### ISSUE THREAD

**Issue by jperals** [samcday/node-google-spreadsheets#20]
When trying to use node-google-spreadsheets on a newly created spreadsheet, I get

```
TypeError: Cannot read property 'feed' of undefined
    at Request._callback ([...]/node_modules/google-spreadsheets/lib/spreadsheets.js:56:16)
    at Request.self.callback ([...]/node_modules/google-spreadsheets/node_modules/request/request.js:372:22)
    at Request.emit (events.js:98:17)
    at Request.<anonymous> ([...]/node_modules/google-spreadsheets/node_modules/request/request.js:1317:14)
    at Request.emit (events.js:117:20)
    at IncomingMessage.<anonymous> ([...]/node_modules/google-spreadsheets/node_modules/request/request.js:1265:12)
    at IncomingMessage.emit (events.js:117:20)
    at _stream_readable.js:929:16
    at process._tickCallback (node.js:419:13)
```

While when using it on an old spreadsheet, it works.

Copying the old spreadsheet and replacing its content manually instead of creating a new spreadsheet seems to make it work, too. But now Google adds a warning on top of the old-format spreadsheets saying "Your older spreadsheets will soon be upgraded to the new version of Google Sheets". So we can't rely on this workaround, of course...

**Comment by samcday** [comment]
@jperals / @msoledade interesting. I haven't had a chance to look into this new Spreadsheet format yet.

Have either of you looked at the API results and seen what is different? Any possibility you can submit a PR to support new spreadsheet format?

**Comment by scottwio** [comment]
Seems to be the ?alt=json doesn't work anymore, try it in the browser you just get an empty document. You can still get xml through the api but not json, odd.

**Comment by jimpick** [comment]
I'm seeing this too.

Here's what Google wrote about the new Google Sheets:

https://support.google.com/docs/answer/6082736

"APIs — You may notice some subtle differences in API behavior."

I was able to use Google's OAuth playground to access the API:

https://developers.google.com/oauthplayground/

I noticed that the requests it generates have a different Authorization header, eg:

Authorization: Bearer ya29.IAGKs4jFzp88Zc1830b....

Currently, the library generates an Authorization header that looks like this:

Authorization: GoogleLogin auth=afadsfasdfasdfa...

(I was using the googleclientlogin npm module to generate a token from a username/password, which still works for old style spreadsheets, but not for the new ones)

I haven't figured out how to set up an app and generate an OAuth2 token that works with the Spreadsheet API outside of the playground yet. I can get tokens working for Google+ using Google's library, but I'm doing something wrong for spreadsheets, or perhaps they just don't allow access. Very frustrating.

**Comment by jimpick** [comment]
Update: I got things working. The short story is that Google is only allowing OAuth2. OAuth1 is deprecated.

https://developers.google.com/google-apps/spreadsheets/#authorizing_requests_with_oauth_20

"Important: OAuth 1.0 is no longer supported and will be disabled on May 5, 2015. If your application uses OAuth 1.0, you must migrate to OAuth 2.0 or your application will cease functioning."

I guess they've jumped the gun by deprecating OAuth1 on new format spreadsheets.

I finally managed to get the 'scopes' right when generating my OAuth2 tokens, and it works for me now.

I'll create a pull request with the change to the Authorization header, as that is needed to support OAuth2.

**Comment by samcday** [comment]
@jimpick great find, thanks!

I've done some reading, and you're right, GoogleClientAuth looks pretty dead at this point. Google expects consumers to be using JWT / OAuth2 (which is quite reasonable in todays security landscape ;)).

I've reworked this library a bit to support the newer styles of Google authentication. When I originally built this library, I deliberately avoided any kind of auth handling code, instead recommending users pull in another library to handle that for them. I definitely think this was the right call, especially after tonight when I discovered that Google have an official Node.js API that can handle oauth2 / JWT authentication + authorisation :)

I've just released version `0.4.0` of this module, which supports auth providers from `googleapis`. Please check it out and let me know what you think.

**Comment by jimpick** [comment]
Thanks for the new release. Everything works now!
